File: frontend/src/effects/decoherence/decoherence.js
import QuantumCircuit from 'quantum-circuit';
import * as utils from '../../utils/imageUtils.js';

/**
 * Simplified Decoherence effect:
 * Computes a quantum decoherence factor for the average pixel of a patch
 * and applies that factor to the entire patch.
 */

const T1_VALUES = [
    0.0002517, // R
    0.0003629, // G
    0.0003598  // B
];

/**
 * Computes a retention factor for a color channel using a quantum circuit.
 * @param {number} val - Average color value [0, 1]
 * @param {number} intensity - User decoherence intensity
 * @param {number} channelIdx - R=0, G=1, B=2
 */
function getDecoherenceFactor(val, intensity, channelIdx) {
    if (val <= 0.001) return 1.0;

    // Use two qubits for Amplitude Damping (monotonic decay)
    const qc = new QuantumCircuit(2);

    // 1. Encode value into qubit 0 (System)
    const theta_in = 2 * Math.asin(Math.sqrt(val));
    qc.addGate("ry", 0, 0, { params: { theta: theta_in } });

    // 2. Amplitude Damping (Decoherence)
    if (intensity > 0) {
        // Calculate decay parameter gamma = 1 - exp(-t/T1)
        // Adjust scale factor to make the 1-20 slider range meaningful
        const t1 = T1_VALUES[channelIdx] || 0.0003;
        const scale = 0.00005;
        const gamma = 1.0 - Math.exp(-(intensity * scale) / t1);

        // theta_noise = 2 * arcsin(sqrt(gamma))
        const theta_noise = 2 * Math.asin(Math.sqrt(Math.max(0, Math.min(1, gamma))));

        // Controlled RY from system (0) to ancilla (1)
        // Decomposition of CRY:
        qc.addGate("ry", 1, 1, { params: { theta: theta_noise / 2 } });
        qc.addGate("cx", 2, [0, 1]);
        qc.addGate("ry", 3, 1, { params: { theta: -theta_noise / 2 } });
        qc.addGate("cx", 4, [0, 1]);

        // CX from ancilla (1) back to system (0)
        qc.addGate("cx", 5, [1, 0]);
    }

    qc.run();

    // 3. Extract the probability of system qubit (q0) being in state |1>
    // In a 2-qubit system (q1, q0), state |1> corresponds to indices 1 (|01>) and 3 (|11>)
    let prob1 = 0;
    const s1 = qc.state[1];
    const s3 = qc.state[3];
    if (s1) prob1 += (s1.re * s1.re + s1.im * s1.im);
    if (s3) prob1 += (s3.re * s3.re + s3.im * s3.im);

    return Math.max(0, Math.min(1.0, prob1 / val));
}

export async function run(params, reportProgress) {
    const image = params.stroke_input.image_rgba;
    const transparentLayer = {
        width: image.width,
        height: image.height,
        data: new Uint8ClampedArray(image.width * image.height * 4),
    };
    const path = params.stroke_input.path;
    const radius = params.user_input.Radius;
    const intensity = params.user_input.Decoherence;

    const border = [image.height, image.width];
    const numSteps = path.length;

    if (reportProgress) reportProgress(0.1);

    for (let i = 0; i < numSteps; i++) {
        if (i % 5 === 0 && reportProgress) {
            reportProgress(0.1 + 0.9 * (i / numSteps));
        }

        const point = [path[i]];
        const region = utils.pointsWithinRadius(point, radius, border);
        if (region.length === 0) continue;

        const patch = utils.getPatch(image, region);

        // Calculate average color for the patch
        let sumR = 0, sumG = 0, sumB = 0;
        for (let k = 0; k < region.length; k++) {
            sumR += patch[k * 4];
            sumG += patch[k * 4 + 1];
            sumB += patch[k * 4 + 2];
        }
        const avgR = sumR / region.length;
        const avgG = sumG / region.length;
        const avgB = sumB / region.length;

        // Compute the quantum decoherence factor based on the average
        const factorR = getDecoherenceFactor(avgR, intensity, 0);
        const factorG = getDecoherenceFactor(avgG, intensity, 1);
        const factorB = getDecoherenceFactor(avgB, intensity, 2);

        const newPatch = new Float32Array(region.length * 4);
        for (let k = 0; k < region.length; k++) {
            newPatch[k * 4] = patch[k * 4] * factorR;
            newPatch[k * 4 + 1] = patch[k * 4 + 1] * factorG;
            newPatch[k * 4 + 2] = patch[k * 4 + 2] * factorB;
            newPatch[k * 4 + 3] = patch[k * 4 + 3]; // Preserve alpha
        }

        utils.setPatch(transparentLayer, region, newPatch);
    }

    if (reportProgress) reportProgress(1.0);
    return transparentLayer;
}

File: backend/effects/decoherence/decoherence.py
import numpy as np
from qiskit import QuantumCircuit
from qiskit.quantum_info import Statevector
from backend import utils

# Decoherence effect (Qiskit version):
# Computes a quantum decoherence factor for the average pixel of a patch
# and applies that factor to the entire patch.

T1_VALUES = [
    0.0002517, # R
    0.0003629, # G
    0.0003598  # B
]

def get_decoherence_factor(val, intensity, channel_idx):
    if val <= 0.001:
        return 1.0
        
    qc = QuantumCircuit(2)
    
    # 1. Encode value into qubit 0 (System)
    theta_in = 2 * np.arcsin(np.sqrt(val))
    qc.ry(theta_in, 0)
    
    # 2. Amplitude Damping (Decoherence)
    if intensity > 0:
        t1 = T1_VALUES[channel_idx] if channel_idx < len(T1_VALUES) else 0.0003
        scale = 0.00005
        gamma = 1.0 - np.exp(-(intensity * scale) / t1)
        
        theta_noise = 2 * np.arcsin(np.sqrt(np.clip(gamma, 0, 1)))
        
        # Controlled RY from 0 to 1
        qc.cry(theta_noise, 0, 1)
        # Reset system if ancilla flipped
        qc.cx(1, 0)
        
    # 3. Extract the probability of system qubit (0) being in state |1>
    sv = Statevector.from_instruction(qc)
    probs = sv.probabilities()
    
    # Probs indices: 00=0, 01=1, 10=2, 11=3.
    # We want qubit 0 to be 1: indices 1 and 3.
    prob1 = probs[1] + probs[3]
    
    return float(np.clip(prob1 / val, 0, 1))

def run(params):
    """
    Executes the Decoherence quantum effect pipeline.
    """
    image = params["stroke_input"]["image_rgba"].copy()
    height, width = image.shape[0], image.shape[1]
    
    # Create the transparent layer we will draw onto
    new_layer = np.zeros_like(image, dtype=np.uint8)
    path = params["stroke_input"]["path"]
    clicks = params["stroke_input"]["clicks"]
    
    # Optional path smoothing interpolation
    if params.get("flags", {}).get("smooth_path", True):
        split_paths = utils.split_path_from_clicks(path, clicks)
        path = np.vstack(split_paths) if split_paths else np.array([])
        
    radius = params["user_input"]["Radius"]
    intensity = params["user_input"]["Decoherence"]

    # Process in sections along the path to avoid too many circuit runs
    # Similar to how it was done in JS
    num_steps = len(path)
    # We'll step through the path
    step_size = max(1, int(radius / 2))
    
    for i in range(0, num_steps, step_size):
        # Current segment of path
        p_segment = path[i : i + step_size]
        region = utils.points_within_radius(p_segment, radius, border=(height, width))
        if len(region) == 0:
            continue
            
        # Get patch and normalize
        patch = image[region[:, 0], region[:, 1]].astype(np.float32) / 255.0
        
        # Average color for the patch
        avg_rgb = np.mean(patch[:, :3], axis=0)
        
        # Start of Validated Algorithm
        # Compute factors
        factors = [get_decoherence_factor(avg_rgb[c], intensity, c) for c in range(3)]
        
        # Apply factors
        new_patch = patch.copy()
        for c in range(3):
            new_patch[:, c] *= factors[c]
        # End of Validated Algorithm
            
        # Re-apply to the transparent layer using utils (handles blending/alpha)
        # We pull the original alpha to keep it unchanged outside of brush
        alpha_patch = image[region[:, 0], region[:, 1], 3:4]
        
        patched_colors = utils.apply_patch_to_image(image[region[:, 0], region[:, 1]], new_patch)
        
        new_layer[region[:, 0], region[:, 1], :3] = patched_colors[:, :3]
        new_layer[region[:, 0], region[:, 1], 3:4] = alpha_patch
        
    return new_layer
